File: core/keyboard_monitor.py
import asyncio
import logging
import time
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Any

import win32api
import win32con


@dataclass
class HotkeyHandler:
    """
    Represents a registered hotkey handler configuration.
    """
    callback: Callable[..., Any]
    cooldown: float
    pass_mouse_pos: bool
    modifiers: List[int]
    last_trigger: float = 0.0


class KeyboardMonitor:
    """
    Monitors keyboard and mouse inputs to trigger registered callbacks.
    
    It supports:
    - Modifier keys (CTRL, SHIFT, ALT)
    - Mouse buttons (Left, Right, Middle, X1, X2)
    - Cooldown management per hotkey
    - Single-press triggering (no spam on hold)
    """

    def __init__(self, logger: logging.Logger):
        """
        Initializes the KeyboardMonitor.

        Args:
            logger: The application logger.
        """
        self.logger: logging.Logger = logger
        self.monitoring: bool = False
        
        # Structure: vk_code -> List of registered handlers for that key
        # We use a list to support multiple combos for the same main key (e.g. F1 vs CTRL+F1)
        self.hotkeys: Dict[int, List[HotkeyHandler]] = {}
        
        # Tracks the "pressed" state of each main key to prevent repeat triggering
        self.key_states: Dict[int, bool] = {}

        self.key_mapping: Dict[str, int] = self._build_key_map()
        self.mouse_mapping: Dict[str, int] = {
            'LEFT': win32con.VK_LBUTTON,
            'RIGHT': win32con.VK_RBUTTON,
            'MIDDLE': win32con.VK_MBUTTON,
            'X1': win32con.VK_XBUTTON1,
            'X2': win32con.VK_XBUTTON2,
        }
        self.modifier_mapping: Dict[str, int] = {
            'CTRL': win32con.VK_CONTROL,
            'SHIFT': win32con.VK_SHIFT,
            'ALT': win32con.VK_MENU
        }

    def _build_key_map(self) -> Dict[str, int]:
        """Builds the mapping for function keys."""
        return {
            f'F{i}': getattr(win32con, f'VK_F{i}')
            for i in range(1, 13)
        }

    def register_hotkey(
        self, 
        key_combo: str, 
        callback: Callable[..., Any], 
        cooldown: float = 0.5, 
        pass_mouse_pos: bool = False
    ) -> None:
        """
        Registers a new hotkey with an associated callback.

        Args:
            key_combo: The key combination string (e.g., "F1", "CTRL+X2").
            callback: The function to call when triggered.
            cooldown: Minimum time in seconds between triggers.
            pass_mouse_pos: If True, passes (x, y) tuple to the callback.
        """
        if not key_combo:
            return
        
        parts = [p.strip().upper() for p in key_combo.split('+')]
        main_key = parts[-1]
        modifiers_str = parts[:-1]

        vk_code = self.key_mapping.get(main_key) or self.mouse_mapping.get(main_key)
        
        if not vk_code:
            self.logger.error(f"Hotkey main key '{main_key}' is not supported in combo '{key_combo}'.")
            return

        required_modifiers = []
        for mod in modifiers_str:
            mod_code = self.modifier_mapping.get(mod)
            if mod_code:
                required_modifiers.append(mod_code)
            else:
                self.logger.warning(f"Unknown modifier '{mod}' in '{key_combo}'")

        if vk_code not in self.hotkeys:
            self.hotkeys[vk_code] = []
            self.key_states[vk_code] = False

        handler = HotkeyHandler(
            callback=callback,
            cooldown=cooldown,
            pass_mouse_pos=pass_mouse_pos,
            modifiers=required_modifiers
        )

        self.hotkeys[vk_code].append(handler)
        
        # Sort handlers by number of modifiers descending.
        # This ensures specific combos (CTRL+X2) are checked before generic ones (X2).
        self.hotkeys[vk_code].sort(key=lambda x: len(x.modifiers), reverse=True)
        
        self.logger.debug(
            f"Registered hotkey '{key_combo}'. Handlers for VK {vk_code}: {len(self.hotkeys[vk_code])}"
        )

    def _is_key_pressed(self, vk_code: int) -> bool:
        """Checks if a virtual key is currently pressed."""
        return win32api.GetAsyncKeyState(vk_code) & 0x8000 != 0

    def _are_modifiers_pressed(self, modifiers: List[int]) -> bool:
        """Checks if all required modifier keys are pressed."""
        for mod_code in modifiers:
            if not self._is_key_pressed(mod_code):
                return False
        return True

    async def start(self) -> None:
        """
        Starts the monitoring loop. Runs indefinitely until stop() is called.
        """
        if not self.hotkeys:
            self.logger.warning("Keyboard monitor started, but no hotkeys are registered.")
            return

        self.monitoring = True
        self.logger.info("Keyboard/Mouse monitor started.")
        poll_interval = 0.02

        while self.monitoring:
            now = time.time()
            
            for vk_code, handlers in self.hotkeys.items():
                is_pressed = self._is_key_pressed(vk_code)
                was_pressed = self.key_states[vk_code]

                # Trigger only on Rising Edge (Released -> Pressed)
                if is_pressed and not was_pressed:
                    matched_handler: Any = None # Typed as Any to avoid strict optional check complexity in loop
                    
                    for handler in handlers:
                        if self._are_modifiers_pressed(handler.modifiers):
                            if (now - handler.last_trigger) > handler.cooldown:
                                matched_handler = handler
                                break
                    
                    if matched_handler:
                        matched_handler.last_trigger = now
                        self._trigger_callback(matched_handler)

                # Update state for next iteration
                self.key_states[vk_code] = is_pressed

            await asyncio.sleep(poll_interval)

    def _trigger_callback(self, handler: HotkeyHandler) -> None:
        """Executes the callback safely."""
        try:
            if handler.pass_mouse_pos:
                mouse_pos = win32api.GetCursorPos()
                if asyncio.iscoroutinefunction(handler.callback):
                    asyncio.create_task(handler.callback(mouse_pos))
                else:
                    handler.callback(mouse_pos)
            else:
                if asyncio.iscoroutinefunction(handler.callback):
                    asyncio.create_task(handler.callback())
                else:
                    handler.callback()
        except Exception as e:
            self.logger.error(f"Error in hotkey callback: {e}", exc_info=True)

    def stop(self) -> None:
        """Stops the monitoring loop."""
        self.monitoring = False
        self.logger.info("Keyboard monitor stopped.")

File: src/main/java/fr/minobot/core/KeyboardMonitor.java
package fr.minobot.core;

import fr.minobot.win32.Point;
import fr.minobot.win32.Win32;
import fr.minobot.win32.WindowApi;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;

/**
 * Polls the keyboard and the mouse buttons, and fires the registered hotkeys — the counterpart of
 * {@code keyboard_monitor.py}.
 *
 * <p>The polling loop runs on a dedicated <em>platform</em> thread, not a virtual one: it is a 50 Hz
 * real-time loop made of native calls, exactly the case where a virtual thread buys nothing and
 * pinning its carrier would be a trap. The callbacks, on the other hand, each get a virtual thread,
 * which is what {@code asyncio.create_task} did — a long {@code invite_all} must not deafen the
 * other hotkeys.
 */
public final class KeyboardMonitor {

    private static final Logger log = LoggerFactory.getLogger(KeyboardMonitor.class);

    private static final Duration POLL_INTERVAL = Duration.ofMillis(20);

    /** The keys whose use means "the user is typing", for {@link FocusManager}'s smart focus. */
    private static final List<Integer> TYPING_KEYS = typingKeys();

    private static final Map<String, Integer> MAIN_KEYS = mainKeys();
    private static final Map<String, Integer> MODIFIER_KEYS = Map.of(
            "CTRL", Win32.VK_CONTROL,
            "SHIFT", Win32.VK_SHIFT,
            "ALT", Win32.VK_MENU);

    private final WindowApi api;
    private final ExecutorService callbackExecutor = Executors.newVirtualThreadPerTaskExecutor();

    /** Virtual key code -> its handlers, most specific (most modifiers) first. */
    private final Map<Integer, List<Hotkey>> hotkeys = new LinkedHashMap<>();
    private final Map<Integer, Boolean> keyStates = new HashMap<>();

    private final AtomicReference<Thread> thread = new AtomicReference<>();

    private volatile boolean monitoring;
    private volatile long lastTypingNanos;
    private volatile boolean everTyped;

    public KeyboardMonitor(WindowApi api) {
        this.api = api;
    }

    /**
     * Registers a hotkey such as {@code "F8"}, {@code "x2"} or {@code "shift+x1"}.
     *
     * <p>An unknown key is logged and ignored rather than fatal, so one bad line in {@code
     * config.json} cannot take the application down.
     */
    public void registerHotkey(String combination, Duration cooldown, Runnable action) {
        register(combination, cooldown, false, cursor -> action.run());
    }

    /** Same, but the action receives the cursor position captured at the moment the key went down. */
    public void registerHotkeyWithCursor(String combination, Duration cooldown, Consumer<Point> action) {
        register(combination, cooldown, true, action);
    }

    private void register(String combination, Duration cooldown, boolean needsCursor, Consumer<Point> action) {
        if (combination == null || combination.isBlank()) {
            return;
        }

        final var parts = combination.split("\\+");
        final var mainKey = parts[parts.length - 1].strip().toUpperCase(Locale.ROOT);

        final var virtualKey = MAIN_KEYS.get(mainKey);
        if (virtualKey == null) {
            log.error("Hotkey main key '{}' is not supported.", mainKey);
            return;
        }

        final var modifiers = new ArrayList<Integer>();
        for (var i = 0; i < parts.length - 1; i++) {
            final var name = parts[i].strip().toUpperCase(Locale.ROOT);
            final var modifier = MODIFIER_KEYS.get(name);
            if (modifier == null) {
                log.warn("Unknown modifier '{}' in '{}'", name, combination);
                continue;
            }
            modifiers.add(modifier);
        }

        final var handlers = hotkeys.computeIfAbsent(virtualKey, key -> new ArrayList<>());
        handlers.add(new Hotkey(combination, List.copyOf(modifiers), cooldown.toNanos(), needsCursor, action));
        // Most specific first: "shift+x2" must win over "x2" when both are registered on X2.
        handlers.sort(Comparator.comparingInt((Hotkey hotkey) -> hotkey.modifiers.size()).reversed());
        keyStates.putIfAbsent(virtualKey, false);

        log.debug("Registered hotkey '{}'.", combination);
    }

    /** Whether the user typed within the given window — the basis of the smart focus. */
    public boolean typedWithin(Duration window) {
        if (!everTyped) {
            return false;
        }
        return System.nanoTime() - lastTypingNanos < window.toNanos();
    }

    /** Starts the polling loop on its own daemon thread and returns immediately. */
    public void start() {
        if (monitoring) {
            log.warn("Keyboard monitor is already running.");
            return;
        }
        monitoring = true;
        thread.set(Thread.ofPlatform()
                .name("keyboard-monitor")
                .daemon(true)
                .start(this::poll));
        log.info("Keyboard/Mouse monitor started.");
    }

    public void stop() {
        if (!monitoring) {
            return;
        }
        monitoring = false;

        final var current = thread.getAndSet(null);
        if (current != null) {
            current.interrupt();
        }
        callbackExecutor.shutdownNow();
        log.info("Keyboard monitor stopped.");
    }

    private void poll() {
        while (monitoring) {
            try {
                long now = System.nanoTime();
                detectTyping(now);
                fireHotkeys(now);

                Thread.sleep(POLL_INTERVAL);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            } catch (RuntimeException e) {
                // A stray failure must not kill the loop: every hotkey would go silent for good.
                log.error("Error in the keyboard polling loop.", e);
            }
        }
    }

    private void detectTyping(long now) {
        for (final var key : TYPING_KEYS) {
            if (api.isKeyDown(key)) {
                lastTypingNanos = now;
                everTyped = true;
                return;
            }
        }
    }

    private void fireHotkeys(long now) {
        for (final var entry : hotkeys.entrySet()) {
            final var virtualKey = entry.getKey();

            final var isDown = api.isKeyDown(virtualKey);
            final var wasDown = keyStates.getOrDefault(virtualKey, false);
            keyStates.put(virtualKey, isDown);

            if (!isDown || wasDown) {
                continue; // hotkeys fire on the edge, not while the key is held
            }

            for (final var hotkey : entry.getValue()) {
                if (!modifiersDown(hotkey.modifiers)) {
                    continue;
                }
                // The first handler whose modifiers match owns this key press, cooldown or not.
                if (hotkey.isReady(now)) {
                    hotkey.markTriggered(now);
                    trigger(hotkey);
                }
                break;
            }
        }
    }

    private boolean modifiersDown(List<Integer> modifiers) {
        return modifiers.stream().allMatch(api::isKeyDown);
    }

    private void trigger(Hotkey hotkey) {
        final var cursor = hotkey.needsCursor ? api.cursorPosition().orElse(null) : null;
        if (hotkey.needsCursor && cursor == null) {
            log.warn("Cannot read the cursor position; '{}' skipped.", hotkey.combination);
            return;
        }

        callbackExecutor.execute(() -> {
            try {
                hotkey.action.accept(cursor);
            } catch (Exception e) {
                log.error("Error in the '{}' hotkey callback.", hotkey.combination, e);
            }
        });
    }

    /** One registered hotkey. Its trigger time is only ever touched by the polling thread. */
    private static final class Hotkey {
        private final String combination;
        private final List<Integer> modifiers;
        private final long cooldownNanos;
        private final boolean needsCursor;
        private final Consumer<Point> action;

        private long lastTriggerNanos;
        private boolean everTriggered;

        Hotkey(String combination, List<Integer> modifiers, long cooldownNanos,
               boolean needsCursor, Consumer<Point> action) {
            this.combination = combination;
            this.modifiers = modifiers;
            this.cooldownNanos = cooldownNanos;
            this.needsCursor = needsCursor;
            this.action = action;
        }

        boolean isReady(long now) {
            return !everTriggered || now - lastTriggerNanos > cooldownNanos;
        }

        void markTriggered(long now) {
            lastTriggerNanos = now;
            everTriggered = true;
        }
    }

    private static Map<String, Integer> mainKeys() {
        final var keys = new LinkedHashMap<String, Integer>();
        for (var i = 1; i <= 12; i++) {
            keys.put("F" + i, Win32.functionKey(i));
        }
        keys.put("LEFT", Win32.VK_LBUTTON);
        keys.put("RIGHT", Win32.VK_RBUTTON);
        keys.put("MIDDLE", Win32.VK_MBUTTON);
        keys.put("X1", Win32.VK_XBUTTON1);
        keys.put("X2", Win32.VK_XBUTTON2);
        return Map.copyOf(keys);
    }

    private static List<Integer> typingKeys() {
        final var keys = new ArrayList<Integer>();
        for (var key = Win32.VK_0; key <= Win32.VK_9; key++) {
            keys.add(key);
        }
        for (var key = Win32.VK_A; key <= Win32.VK_Z; key++) {
            keys.add(key);
        }
        keys.add(Win32.VK_SPACE);
        keys.add(Win32.VK_RETURN);
        keys.add(Win32.VK_BACK);
        keys.add(Win32.VK_TAB);
        return List.copyOf(keys);
    }
}
